sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -309,53 +308,103 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
-                                          InsertSourceGen insertSourceGen,
-                                          @Nullable ReturnValueGen returnGen,
-                                          boolean isRetry) throws Exception {
-        final long seqNo;
-        final long primaryTerm;
-        final long version;
-        Doc updatedDoc = null;
-        if (tryInsertFirst) {
-            version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE
-                ? Versions.MATCH_ANY
-                : Versions.MATCH_DELETED;
-            seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
-            primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
-            try {
-                item.source(insertSourceGen.generateSourceAndCheckConstraints(item.insertValues()));
-            } catch (IOException e) {
-                throw ExceptionsHelper.convertToElastic(e);
+    protected IndexItemResponse insert(ShardUpsertRequest request,
+                                       ShardUpsertRequest.Item item,
+                                       IndexShard indexShard,
+                                       boolean isRetry,
+                                       @Nullable ReturnValueGen returnGen,
+                                       @Nullable InsertSourceGen insertSourceGen) throws Exception {
+        assert insertSourceGen != null : "InsertSourceGen must not be null";
+        final BytesReference rawSource;
+        Map<String, Object> source = null;
+        try {
+            if (insertSourceGen instanceof FromRawInsertSource) {
+                rawSource = insertSourceGen.generateSourceAndCheckConstraintsAsBytesReference(item.insertValues());
+            } else {
+                source = insertSourceGen.generateSourceAndCheckConstraints(item.insertValues());
+                rawSource = insertSourceGen.toBytesReference(source);
             }
-        } else {
-            Doc currentDoc = getDocument(indexShard, item.id(), item.version(), item.seqNo(), item.primaryTerm());
-
-            Map<String, Object> updatedSource = updateSourceGen.generateSource(
-                currentDoc,
-                item.updateAssignments(),
-                item.insertValues()
-            );
-
-            if (item.returnValues() != null) {
-                updatedDoc = currentDoc.withUpdatedSource(updatedSource);
+        } catch (IOException e) {
+            throw ExceptionsHelper.convertToElastic(e);
+        }
+        item.source(rawSource);
+
+        final long version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE ? Versions.MATCH_ANY : Versions.MATCH_DELETED;
+        final long seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
+        final long primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
+
+        Engine.IndexResult indexResult = index(item, indexShard, isRetry, seqNo, primaryTerm, version);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            if (source == null) {
```

Comment:
According to @mfussenegger this case should never happen since the `source` would be only null if it is created from a `FromRawInsertSource` which only applies to the `COPY` statement.

Reply:
I think I was wrong. a `INSERT INTO tbl (_raw) ...` could probably also trigger this case.